File: todo-api-js/src/adapters/controllers/todo/create.ts
import { type UseCase } from "@domain/use-case";
import { type CreateTodoInput, type CreateTodoOutput } from "@application/dtos/todo/create";
import { type CreateTodoRequest, type CreateTodoPresenter } from "@adapters/dtos/todo/create";

export class CreateTodoController<TPresenter extends CreateTodoPresenter<any>> {
    constructor(
        private interactor: UseCase<CreateTodoInput, CreateTodoOutput>,
        private presenter: TPresenter
    ) {}

    async run(req: CreateTodoRequest): Promise<ReturnType<TPresenter["present"]>> {
        const input = req.parse();
        if (input.isErr()) {
            return this.presenter.present(input);
        }

        const result = await this.interactor.exec(input.value);
        return this.presenter.present(result);
    }
}

File: todo-api-rs/src/adapters/controllers/todo/create.rs
use crate::adapters::dtos::todo::create::{
    CreatePresenter, CreateRequest, CreateResponseError,
};
use crate::application::dtos::todo::create::{CreateTodoError, CreateTodoInput, CreateTodoOutput};
use crate::domain::use_case::UseCase;

#[derive(Debug)]
pub struct CreateTodoController<T, P> {
    interactor: T,
    presenter: P,
}

impl<T, P> CreateTodoController<T, P>
where
    T: UseCase<CreateTodoInput, CreateTodoOutput>,
    P: CreatePresenter,
{
    pub const fn new(interactor: T, presenter: P) -> Self {
        Self {
            interactor,
            presenter,
        }
    }

    pub async fn run(self, req: CreateRequest) -> <P as CreatePresenter>::View {
        let input = match req.parse().map_err(CreateResponseError::Input) {
            Ok(input) => input,
            Err(err) => return self.presenter.present(Err(err)),
        };

        let result = self.interactor.exec(input).await.map_err(|err| match err {
            CreateTodoError::DuplicatedTitle(title) => {
                CreateResponseError::DuplicatedTitle(title)
            }
            CreateTodoError::Internal(src) => CreateResponseError::Internal(src),
        });

        self.presenter.present(result)
    }
}
